web/js/codeworld_shared.js

```diff
@@ -122,95 +122,161 @@ var hintBlacklist = [
-                elem.title = line;
+        }
+        // foo =
+        else if (/^\S+\s*=/.test(line)) {
+            word = line.split("=")[0].trim();
+            parseResults[word] = {
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
+            };
+        }
+        // data Foo
+        else if (/^data\s.+/.test(line)) {
+            word = line.split(" ")[1];
+            parseResults[word] = {
+                declaration: line.slice(0, getWordStart(word, line)),
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
             }
         }
-        return {
-            text: word,
-            render: renderer,
-            source: line
-        };
-    }
-
-    function createHover(line, wordStart, wordEnd, doc, hint){
-        if (!hint){
-            hint = createHint(line, wordStart, wordEnd);
+        // type Foo = Bar
+        else if (/^type\s.+/.test(line)) {
+            var splitted = line.split("=");
+            word = splitted[0].trim().split(" ").slice(-1)[0];
+            parseResults[word] = {
+                declaration: line,
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
+            };
         }
+        // (*#^) :: Type
+        else if (/^\([^\(\)]+\)\s*::/.test(line)) {
+            var splitted = line.split("::");
+            var word = splitted[0].trim()
+            word = word.slice(1, word.length - 1)
+            parseResults[word] = {
+                declaration: line,
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line)
+            }
+        }
+        // foo :: Type
+        else if (/^\S+\s*::/.test(line)) {
+            var splitted = line.split("::");
+            word = splitted[0].trim()
+            parseResults[word] = {
+                declaration: line,
+                symbolStart: getWordStart(word, line),
+                symbolEnd: getWordEnd(word, line),
+            }
+        }
+    })
+    codeWorldSymbols = Object.assign({}, parseResults, codeWorldBuiltinSymbols);
+};
 
-        var topDiv = document.createElement('div');
-
-        var docDiv = document.createElement('div');
-        var annotation = document.createElement('div');
-        hint.render(annotation);
-        annotation.className = "hover-decl";
-        docDiv.appendChild(annotation);
+function renderHover(keyword) {
+    var topDiv = document.createElement('div')
 
-        if (doc !== "") {
-            var description = document.createElement('div');
-            description.innerHTML = doc;
-            description.className = " hover-doc";
-            docDiv.appendChild(description);
-        }
+    if (!codeWorldSymbols[keyword] // no entry
+        || // no docs or declaration - nothing to show on hover
+        !(codeWorldSymbols[keyword].doc
```

Comment:
Why not assume that `declaration` will always be present, and when you're parsing, fill it in with just the symbol itself if you can't do any better?  That simplifies this logic a lot.  In fact, there's already code later that assumes `declaration` is populated.